bin/ch-run.c

```diff
@@ -103,191 +59,81 @@ const struct argp_option options[] = {
    { 0 }
 };
 
-struct bind {
-   char * src;
-   char * dst;
-};
-
 struct args {
-   struct bind binds[USER_BINDS_MAX+1];
-   gid_t container_gid;
-   uid_t container_uid;
-   char * newroot;
+   struct container c;
    char * initial_dir;
-   bool join;
-   int join_ct;
-   char * join_tag;
-   bool private_home;
-   bool private_tmp;
-   bool writable;
-   int user_cmd_start;  // index into argv where NEWROOT is
 };
 
-void enter_udss(char * newroot, bool writeable, struct bind * binds,
-                bool private_tmp, bool private_home);
+/** Function prototypes **/
+
 bool get_first_env(char ** array, char ** name, char ** value);
-void join_begin();
-int join_ct();
-void join_end();
-void join_namespace(pid_t pid, char * ns);
-void join_namespaces();
-char * join_tag();
-void log_ids(const char * func, int line);
-void run_user_command(int argc, char * argv[], int user_cmd_start);
-static error_t parse_opt(int key, char * arg, struct argp_state * state);
+int join_ct(int cli_ct);
+char * join_tag(char * cli_tag);
 int parse_int(char * s, bool extra_ok, char * error_tag);
+static error_t parse_opt(int key, char * arg, struct argp_state * state);
 void privs_verify_invoking();
-void sem_timedwait_relative(sem_t * sem, int timeout);
-void setup_namespaces(uid_t cuid, gid_t cgid);
 
-struct args args;
+
+/** Global variables **/
+
 const struct argp argp = { options, parse_opt, args_doc, usage };
 
 
 /** Main **/
 
 int main(int argc, char * argv[])
 {
+   struct args args;
+   int arg_next;
+   int c_argc;
+   char ** c_argv;
+
    privs_verify_invoking();
-   memset(args.binds, 0, sizeof(args.binds));
-   args.container_gid = getegid();
-   args.container_uid = geteuid();
+
+   T_ (args.c.binds = calloc(1, sizeof(struct bind)));
```

Comment:
nitpick: I remember seeing dialog about renaming TRYX a while back. I also remember being surprised to learn that the Tf_, Te_, Z_  nomenclature is some kind of standard.

Although I agree that TRYX() needed some renaming, I feel like this new syntax is unintuitive and introduces ambiguity. Maybe after opening charliecloud.h a bunch more times it will stick...?

Reply:
Fair enough. The reasoning is "true" versus "zero". Then you get *f* for `printf(2)`-style, *e* for without `errno`, underscore to keep them the same width.... Good feedback that they're still a bit confusing.